### PR TITLE
Add hooks for prefactorization of linear solvers

### PR DESCRIPTION
**Pull request by simonbyrne**
Add a `prefactorize` function that can is applied to linear solvers at setup time. Also changes the argument order to `linearsolve!` to make it consistent.

### CODE REVIEW

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try-

**Comment by simonbyrne** [comment]
bors try

File: src/ODESolvers/AdditiveRungeKuttaMethod.jl [hunk 1]
```diff
@@ -14,6 +14,32 @@ using ..SpaceMethods
 using ..LinearSolvers
 using ..MPIStateArrays: device, realview
 
+
+
+"""
+    op! = EulerOperator(f!, ϵ)
+
+Constuct a linear operator which performs an explicit Euler step ``Q + α f(Q)``,
```
Comment (lcw) [inline src/ODESolvers/AdditiveRungeKuttaMethod.jl]:
```suggestion
Construct a linear operator which performs an explicit Euler step ``Q + α f(Q)``,
```

Comment (jkozdon) [reply src/ODESolvers/AdditiveRungeKuttaMethod.jl]:
```suggestion
Construct a linear operator which performs an explicit Euler step ``Q + ϵ f(Q)``,
```

File: src/ODESolvers/AdditiveRungeKuttaMethod.jl [hunk 2]
```diff
@@ -111,7 +144,13 @@ function AdditiveRungeKutta(spacedisc::AbstractSpaceMethod,
                      Q; dt=dt, t0=t0)
 end
 
-ODEs.updatedt!(ark::AdditiveRungeKutta, dt) = (ark.dt = dt)
+function ODEs.updatedt!(ark::AdditiveRungeKutta, dt)
+  ark.dt = dt
+  # this will only work for implicit solves
```
Comment (lcw) [inline src/ODESolvers/AdditiveRungeKuttaMethod.jl]:
```suggestion
  # this will only work for iterative solves
```

**Comment by jkozdon** [comment]
We may want to throw a warning or error if `adjustfinalstep = true` and the matrix is being factored, otherwise the matrix will be built special for the final time step.

Other options:

- adjust final step removed and just number of steps is specified
- adjust final step behavior changed to uniformly change steps for all of the stepping

**Comment by simonbyrne** [comment]
included as part of #518

**Pull request closed without merging**